src/opset/Gemm_13.c

```diff
@@ -0,0 +1,161 @@
+/*
+ *  CONNX, C implementation of Open Neural Network Exchange Runtime
+ *  Copyright (C) 2019-2022 TSN Lab, Inc.
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU General Public License as published by
+ *  the Free Software Foundation, either version 3 of the License, or
+ *  (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU General Public License for more details.
+ *
+ *  You should have received a copy of the GNU General Public License
+ *  along with this program.  If not, see <https://www.gnu.org/licenses/>.
+ */
+
+#include <assert.h>
+#include <math.h>
+#include <stdio.h>
+
+#include <connx/accel.h>
+#include <connx/connx.h>
+
+// clang-format off
+int Gemm_{{op_version}}(connx_Graph* graph, __attribute__((unused)) uint32_t output_count, uint32_t* outputs,
+                         // clang-format on
+                         __attribute__((unused)) uint32_t input_count, uint32_t* inputs,
+                         __attribute__((unused)) uint32_t attribute_count, __attribute__((unused)) void** attributes) {
+
+    connx_Tensor* A = connx_Graph_get(graph, inputs[0]);
+    connx_Tensor* B = connx_Graph_get(graph, inputs[1]);
+    assert(A != NULL && B != NULL);
+
+    float32_t alpha = *(float32_t*)attributes[0];
+    float32_t beta = *(float32_t*)attributes[1];
+    bool transA = *(int32_t*)attributes[2] ? true : false;
+    bool transB = *(int32_t*)attributes[3] ? true : false;
+
+    float32_t(*a)[A->shape[1]] = A->buffer;
+    float32_t(*b)[B->shape[1]] = B->buffer;
+
+    // alias for A, B rows/cols whether transposed flags enabled
+    int32_t A_rows = A->shape[0];
+    int32_t A_cols = A->shape[1];
+    int32_t B_rows = B->shape[0];
+    int32_t B_cols = B->shape[1];
+
+    if (transA) {
+        A_rows = A->shape[1];
+        A_cols = A->shape[0];
+    }
+
+    if (transB) {
+        B_rows = B->shape[1];
+        B_cols = B->shape[0];
+    }
+
+    // A's # of cols and B's # of row should be same
+    if (A_cols != B_rows) {
+        connx_error("Gemm: A_cols != B_rows");
+        return CONNX_TENSOR_SHAPE_NOT_MATCHING;
+    }
+
+    connx_Tensor* C = NULL;
+    bool is_biased = false;
+    if (input_count == 3) {
+        C = connx_Graph_get(graph, inputs[2]);
+        assert(C != NULL);
+        // Check if Unidirectional Broadcasting is possible,
+        // result shape should be (A_rows, B_cols)
+        if (A_rows == C->shape[0] && B_cols == C->shape[1]) { // same shape
```

Comment:
```
if ndim = 0
  true
elif ndim = 1
  C_shape[0] = 1 or C_shape[0] = B_cols
elif ndim = 2
  C_shape[0] in (1, B_cols) and C_shape[1] in (1, A_rows)
else
  error
```

Reply:
Also, Make `int32_t shape[] = {A_rows, B_cols}` First, Then use `shape[0]` `shape[1]` instead of A_rows, B_cols would be better readability